### ISSUE TITLE
Provide a sample sudoers file for the suggested csi user

### ISSUE THREAD

**Issue by sjmudd** [democratic-csi/democratic-csi#34]
The documentation suggests creating a `csi` user to avoid ssh as a `root` user and to then enable sudo for the required zfs commands.

I see for my own ZoL (on ubuntu) that a `/etc/sudoers.d/zfs` file is created which would probably almost cover the required access, so suggest using this as a template (and minimising required privileges as `help` is not needed) and adding this to the documentation.

e.g. the file I see is:

```
## Allow read-only ZoL commands to be called through sudo
## without a password. Remove the first '#' column to enable.
##
## CAUTION: Any syntax error introduced here will break sudo.
##
## Cmnd alias specification
#Cmnd_Alias C_ZFS = \
#  /sbin/zfs "", /sbin/zfs help *, \
#  /sbin/zfs get, /sbin/zfs get *, \
#  /sbin/zfs list, /sbin/zfs list *, \
#  /sbin/zpool "", /sbin/zpool help *, \
#  /sbin/zpool iostat, /sbin/zpool iostat *, \
#  /sbin/zpool list, /sbin/zpool list *, \
#  /sbin/zpool status, /sbin/zpool status *, \
#  /sbin/zpool upgrade, /sbin/zpool upgrade -v
#
## allow any user to use basic read-only ZFS commands
#ALL ALL = (root) NOPASSWD: C_ZFS
```

So suggesting the creation of a `/etc/sudoers.d/csi` and replacing `zfs` with `csi` and `C_ZFS` with `C_CSI` would probably work but removing the `help`, `iostat` options and adding anything that's missing would be really good to keep the ZFS access as secure as possible.  I'm not familiar enough yet with this driver or ZFS to be able to provide the PR but guess this should be reasonably straightforward.

**Comment by sjmudd** [comment]
Also there is a bit of confusion in the docs.

If using `sudo` to access the zfs server it probably makes sense to clearly differentiate the `csi-provisioner` username, used on the kubernetes "client side" to talk to the ZFS server from the `csi` user on the ZFS server itself.   Using ssh keys means the usernames are clearly different and the ZFS server's unix `csi` account only needs the addition of a `~csi/.ssh/authorized_keys` file with the public key of the `csi-provisioner`'s user.

**Comment by travisghansen** [comment]
Thanks for participating in the project! The documentation certainly could use some help in this area to make it more clear what ‘server’ is being referenced etc.

I’ll need to do a little research around the sudoers file and what commands are exactly required. It’s pretty minimal *but* there are scenarios where I execute a shell as sudo (when enabled) which makes it pretty useless to attempt to limit what the user can execute via sudo (since if you add `sh` for example that implies everything anyhow). Shell is required in some very specific scenarios because I need to use pipes and things of that nature. There could be ways around that though not sure.

**Comment by sjmudd** [comment]
That's a valid point and if made/explained then people can decide if they want to use a tighter sudo configuration.  Either way personally I prefer not to allow direct ssh access as root and know this is not generally considered "good", so the use of a different account to use with ssh seems better, even if the subsequent sudo access allows full access after gaining access to the server.

**Comment by Elegant996** [comment]
It's now possible to use `sudo` with `nopasswd` and to specify commands in TrueNAS 12.0-U3. Looks like the web interface doesn't support it yet but you can make the modifications directly in the config:

```
root@truenas[~]# sqlite3 /data/freenas-v1.db
sqlite> UPDATE account_bsdgroups SET bsdgrp_sudo_nopasswd='1' WHERE bsdgrp_gid='1001';
sqlite> .q
```

In the above case, user `csi` is a member of group `1001`. Once the change has been made, reload the web interface and you will see the change by running `cat /etc/local/sudoers`.

We can further customize this by specifying a json for `bsdgrp_sudo_commands` but would need the list of commands. Would it be possible to list which cases `sh` is required so we could seek to secure the `csi` account's access?

Thanks!

**Comment by travisghansen** [comment]
I’ll send some commands in a bit but there is a more proper way to do this using the fancy new cli tool.

**Comment by travisghansen** [comment]
OK, so with `12.0-u3` this is the proper way to do this:

```
# at the command promp
cli

# after you enter the truenas cli and are at that prompt
account user query select=id,username,uid,sudo_nopasswd

# find the `id` of the user you want to update (note, this is distinct from the `uid`)
account user update id=<id> sudo_nopasswd=true

# exit cli by hitting ctrl-d

# confirm sudoers file is appropriate
cat /usr/local/etc/sudoers
```

**Comment by Elegant996** [comment]
This would provide the `csi` user with root level access via `sudo`. Is it possible to restrict the commands required? There was mention that `sh` was required before, does this still hold true? So long as `sh` is required, restricting sudo commands won't really make any impact.

**Comment by travisghansen** [comment]
I would need to go audit it all, but I'm fairly confident the only time `sh` is used to for 'detached' snapshots. In that case `sh` is used instead of the commands directly because I do a `zfs send | zfs receive`. It may be possible to mimic a `|` in pure js but I haven't looked into it deeply.

The struggle beyond that will be getting support in the TrueNAS platform configuration to put in a list of allowed commands. I'm not sure I have the answer here of how to 'solve' these problems completely. To me if you look at the realities of the situation more broadly, if the user can create/delete storage (on a system dedicated to storage), what extra danger is there from having `sh`? I ask that not rhetorically...I'm wondering if we can address the other items fundamentally or not. I would also note, the same holds true for the api...the api user can create/delete storage, reboot the node, trigger a software update, etc...for a storage system what can be worse than deleting storage and/or just shutting down the machine?

I don't know, I love the discussion and feedback though as I feel we'll ultimately make it better as we iterate.

**Comment by travisghansen** [comment]
Documented how to setup the sudoers file in the README

**Comment by travisghansen** [comment]
Note that an experimental driver has been released which can be used with SCALE 21.08+ and v1.3.0 of democratic-csi+ that removes the need for ssh entirely. All operations are performed over the TrueNAS api.
